File: ocaml/get.pl
#! /usr/bin/perl
#
# You have entered a maze of twisty passages, all alike.

use 5.008;
use strict;

die "Expecting an argument" unless $#ARGV == 0;
die "Expecting a single integer" unless $ARGV[0] =~ /([1-9][0-9]*)/;

my $problem = int($1);
my $prob3 = sprintf("%03d", $problem);
my $URL = "http://projecteuler.net/problem=$problem";
open(I, "w3m -dump -cols 69 $URL|") or die;

open(O, ">pr$prob3.ml") or die;
print O <<EOF;
(**********************************************************************
EOF

# Skip until we get the problem statement.
while (<I>) {
	last if /^Problem /;
}
print O " * $_";
my $last = $_;
while (<I>) {
	last if /Project Euler Copyright/;
	if ($_ ne $last || !/^\s*\n/) {
		chomp;
		my $line = " * $_";
		$line =~ s/\s+$//;
		print O  "$line\n";
	}
	$last = $_;
}
print O <<EOF;
 **********************************************************************)

EOF
close I or die "Problem getting euler page";
close O or die;

File: erlang/get.py
#! /usr/bin/env python3

# This seems fairly obvious to implement this in Python for this
# particular exercise.

import os
import subprocess
import sys

def generate(problem, contents):
    name = 'pr{0:03}.erl'.format(problem)
    with open(name, 'x') as f:
        for line in contents:
            if len(line) > 0:
                print("%", line.decode(), file=f)
            else:
                print("%", file=f)
        print('''
-module(pr{0:03}).
-export([solve/0]).

solve() ->
	42. '''.format(problem), file=f)

        # os.chmod(name, 0o755)

def main():
    if len(sys.argv) != 2:
        print("Expecting a single argument")
        sys.exit(1)
    problem = int(sys.argv[1])
    p3 = '{0:03}'.format(problem)
    url = '../haskell/probs/problem-{0:03}.html'.format(problem)
    contents = subprocess.check_output(['w3m', '-dump', '-cols', '75',
        url]).splitlines()

    # Skip until we get to the problem statement.
    while not contents[0].startswith(b'Problem '):
        del contents[0]

    # Remove everything after the copyright line.
    for i in range(len(contents)):
        if contents[i].startswith(b'Project Euler Copyright'):
            del contents[i:]
            break

    # Remove blanks at the end.
    while contents[-1] == b'':
        del contents[-1]

    generate(problem, contents)

if __name__ == '__main__':
    main()
